### PR TITLE
Wrap async callbacks in try catch

### LINKED ISSUE #3891: Wrap async callbacks in try / catch
Currently, if async callbacks throw exceptions then the collect process fails for all instruments. Async callbacks are not wrapped in a try / catch but they should be. If an exception occurs, we should log a diagnostic warning message.

### PR DESCRIPTION
**Pull request by jack-berg**
Resolves #3891.

### CODE REVIEW

**Review by jsuereth — APPROVED** [review]

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AsynchronousMetricStorage.java [hunk 1]
```diff
@@ -143,7 +143,17 @@ public MetricData collectAndReset(
...
+      try {
+        metricUpdater.run();
+      } catch (RuntimeException e) {
+        logger.log(
+            Level.WARNING,
+            "An exception occurred invoking callback for instrument "
+                + getMetricDescriptor().getName()
+                + ".",
+            e);
+        return null;
```
Comment (jack-berg) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AsynchronousMetricStorage.java:155]: This should return `MetricData.empty()` once `@Nullable` annotation is removed in #3887.

**Review by jkwatson — APPROVED** [review]

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AsynchronousMetricStorage.java [hunk 2]
```diff
@@ -143,7 +143,17 @@ public MetricData collectAndReset(
             collectionInfo.getSupportedAggregation(), collectionInfo.getPreferredAggregation());
     collectLock.lock();
     try {
-      metricUpdater.run();
+      try {
+        metricUpdater.run();
+      } catch (RuntimeException e) {
```
[the remark this answers is not in the document]
Comment (jkwatson) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AsynchronousMetricStorage.java:148]: But the API we provide doesn't allow for throwing checked exceptions, does it?

[the remark this answers is not in the document]
Comment (anuraaga) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AsynchronousMetricStorage.java:148]: `Error` isn't checked, nor is it `RuntimeException`. Quite silly design of the JDK :)

[the remark this answers is not in the document]
Comment (jkwatson) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AsynchronousMetricStorage.java:148]: I feel like if an Error is thrown, all bets are off with the behavior of anything. AssertionError being the only exception (heh) to that. I'm not opposed to changing this to Throwable in a followup, though. although, I'd hate to swallow an OOME

[the remark this answers is not in the document]
Comment (anuraaga) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AsynchronousMetricStorage.java:148]: Yeah AssertionError is annoying - I think in practice, non-fatal Errors do tend to crop up and are recoverable vs fatal errors like OOME, which generally indicate you're hosed anyways. So we should switch to Throwable, but of course it doesn't hurt to add the propagateIfFatal helper in the same PR :)

**Pull request merged**